cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3850 @@
+            F(S_NATIONKEY)                      \
+            F(S_PHONE)                          \
+            F(S_ACCTBAL)                        \
+            F(S_COMMENT)
+
+#define MAKE_ENUM(col) col,
+            struct SUPPLIER
+            {
+                enum
+                {
+                    FOR_EACH_COLUMN(MAKE_ENUM)
+                    kNumCols,
+                };
+            };
+#undef MAKE_ENUM
+#define MAKE_STRING_MAP(col)                    \
+            { #col, SUPPLIER::col },
+            const std::unordered_map<std::string, int> kNameMap =
+            {
+                FOR_EACH_COLUMN(MAKE_STRING_MAP)
+            };
+#undef MAKE_STRING_MAP
+#define MAKE_FN_ARRAY(col)                                              \
+            [this](size_t thread_index) { return this->col(thread_index); },
+            std::vector<GenerateColumnFn> kGenerators =
+            {
+                FOR_EACH_COLUMN(MAKE_FN_ARRAY)
+            };
+#undef MAKE_FN_ARRAY
+#undef FOR_EACH_COLUMN
+
+            std::vector<std::shared_ptr<DataType>> kTypes =
+            {
+                int32(),
+                fixed_size_binary(25),
+                utf8(),
+                int32(),
+                fixed_size_binary(15),
+                decimal(12, 2),
+                utf8(),
+            };
+
+            Status ProduceCallback(size_t thread_index)
+            {
+                if(done_.load())
+                    return Status::OK();
+                ThreadLocalData &tld = thread_local_data_[thread_index];
+                tld.suppkey_start = rows_generated_.fetch_add(batch_size_);
+                if(tld.suppkey_start >= rows_to_generate_)
+                    return Status::OK();
+
+                tld.to_generate = std::min(batch_size_,
+                                           rows_to_generate_ - tld.suppkey_start);
+
+                tld.batch.resize(SUPPLIER::kNumCols);
+                std::fill(tld.batch.begin(), tld.batch.end(), Datum());
+                for(int col : gen_list_)
+                    RETURN_NOT_OK(kGenerators[col](thread_index));
+
+                std::vector<Datum> result(gen_list_.size());
+                for(size_t i = 0; i < gen_list_.size(); i++)
+                {
+                    int col_idx = gen_list_[i];
+                    result[i] = tld.batch[col_idx];
+                }
+                ARROW_ASSIGN_OR_RAISE(ExecBatch eb, ExecBatch::Make(std::move(result)));
+                int64_t batches_to_generate = (rows_to_generate_ + batch_size_ - 1) / batch_size_;
+                int64_t batches_outputted_before_this_one = batches_outputted_.fetch_add(1);
+                bool is_last_batch = batches_outputted_before_this_one == (batches_to_generate - 1);
+                output_callback_(std::move(eb));
+                if(is_last_batch)
+                {
+                    done_.store(true);
+                    finished_callback_(batches_outputted_.load());
```

Comment:
We should set `done_` with `compare_exchange_strong` like we do elsewhere and guard the call to `finished_callback_`